velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,452 @@
+  if (constExpr->hasValueVector()) {
+    return toSubstraitLiteral(arena, constExpr->valueVector(), litValue);
+  } else {
+    return toSubstraitLiteral(arena, constExpr->value());
+  }
+}
+
+const ::substrait::Expression_Literal&
+VeloxToSubstraitExprConvertor::toSubstraitLiteral(
+    google::protobuf::Arena& arena,
+    const velox::variant& variantValue) {
+  ::substrait::Expression_Literal* literalExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Literal>(
+          &arena);
+  switch (variantValue.kind()) {
+    case velox::TypeKind::DOUBLE: {
+      literalExpr->set_fp64(variantValue.value<TypeKind::DOUBLE>());
+      break;
+    }
+    case velox::TypeKind::VARCHAR: {
+      ::substrait::Expression_Literal::VarChar* substraitVarChar =
+          google::protobuf::Arena::CreateMessage<
+              ::substrait::Expression_Literal::VarChar>(&arena);
+      auto charValue = variantValue.value<StringView>();
+      substraitVarChar->set_value(charValue.data());
+      substraitVarChar->set_length(charValue.size());
+      literalExpr->set_allocated_var_char(substraitVarChar);
+      break;
+    }
+    case velox::TypeKind::BIGINT: {
+      literalExpr->set_i64(variantValue.value<TypeKind::BIGINT>());
+      break;
+    }
+    case velox::TypeKind::INTEGER: {
+      literalExpr->set_i32(variantValue.value<TypeKind::INTEGER>());
+      break;
+    }
+    case velox::TypeKind::SMALLINT: {
+      literalExpr->set_i16(variantValue.value<TypeKind::INTEGER>());
+      break;
+    }
+    case velox::TypeKind::TINYINT: {
+      literalExpr->set_i8(variantValue.value<TypeKind::INTEGER>());
+      break;
+    }
+    case velox::TypeKind::BOOLEAN: {
+      literalExpr->set_boolean(variantValue.value<TypeKind::BOOLEAN>());
+      break;
+    }
+    case velox::TypeKind::REAL: {
+      literalExpr->set_fp32(variantValue.value<TypeKind::REAL>());
+      break;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      // TODO make sure the type convertor is equal
+      literalExpr->set_timestamp(
+          variantValue.value<TypeKind::TIMESTAMP>().getNanos());
+      break;
+    }
+    default:
+      VELOX_NYI(
+          "Unsupported constant Type '{}' ",
+          mapTypeKindToName(variantValue.kind()));
+  }
+
+  return *literalExpr;
+}
+
+const ::substrait::Expression_Literal&
+VeloxToSubstraitExprConvertor::toSubstraitLiteral(
+    google::protobuf::Arena& arena,
+    const velox::VectorPtr& vectorValue,
+    ::substrait::Expression_Literal_Struct* litValue) {
+  ::substrait::Expression_Literal* substraitField =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Literal>(
+          &arena);
+  const TypePtr& childType = vectorValue->type();
+
+  switch (childType->kind()) {
+    case velox::TypeKind::BOOLEAN: {
+      auto childToFlatVec = vectorValue->asFlatVector<bool>();
+      // Get the batchSize and convert each value in it.
+      vector_size_t flatVecSize = childToFlatVec->size();
```

Comment:
This logic repeats multiple times here. Consider using  VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH to reduce copy-paste.

Reply:
fix.